src/Features/Core/Portable/Completion/Providers/TypeImportCompletionService.cs

```diff
@@ -0,0 +1,357 @@
+            {
+                if (!(compilation.GetAssemblyOrModuleSymbol(peReference) is IAssemblySymbol assemblySymbol))
+                {
+                    return ImmutableArray<TypeImportCompletionItem>.Empty;
+                }
+
+                var key = GetReferenceKey(peReference);
+                var isInternalsVisible = compilation.Assembly.IsSameAssemblyOrHasFriendAccessTo(assemblySymbol);
+                var rootNamespaceSymbol = assemblySymbol.GlobalNamespace;
+
+                if (key == null)
+                {
+                    // Can't cache items for reference with null key, so just create them and return. 
+                    return GetCompletionItemsForTopLevelTypeDeclarations(rootNamespaceSymbol, ns => !excludedNamespaces.Contains(ns), isInternalsVisible);
+                }
+
+                var checksum = SymbolTreeInfo.GetMetadataChecksum(solution, peReference, cancellationToken);
+                return GetAccessibleTopLevelTypesWorker(key, rootNamespaceSymbol, checksum, excludedNamespaces, isInternalsVisible, _peItemsCache, cancellationToken);
+
+                static string GetReferenceKey(PortableExecutableReference reference)
+                    => reference.FilePath ?? reference.Display;
+            }
+
+            private static ImmutableArray<TypeImportCompletionItem> GetAccessibleTopLevelTypesWorker<TKey>(
+                TKey key,
+                INamespaceSymbol rootNamespace,
+                Checksum checksum,
+                ImmutableHashSet<string> excludedNamespaces,
+                bool isInternalsVisible,
+                ConcurrentDictionary<TKey, ReferenceCacheEntry> cache,
+                CancellationToken cancellationToken)
+            {
+                var tick = Environment.TickCount;
+                var returned = ImmutableArray<TypeImportCompletionItem>.Empty;
+                var created = ImmutableArray<TypeImportCompletionItem>.Empty;
+#if DEBUG
+                try
+#endif
+                {
+                    // Cache miss, create all requested items.
+                    if (!cache.TryGetValue(key, out var cacheEntry) ||
+                        cacheEntry.Checksum != checksum ||
+                        !AccessibilityMatch(cacheEntry.IncludeInternalTypes, isInternalsVisible))
+                    {
+                        var items = GetCompletionItemsForTopLevelTypeDeclarations(rootNamespace, ns => !excludedNamespaces.Contains(ns), isInternalsVisible);
+                        cache[key] = new ReferenceCacheEntry(checksum, isInternalsVisible, excludedNamespaces, items);
+
+                        returned = created = items;
+                        return items;
+                    }
+
+                    // Having fewer excluded names space in cache than in request means the cache contains items for all the types not excluded.
+                    if (cacheEntry.ExcludedNamespaces.IsSubsetOf(excludedNamespaces))
```

Comment:
all the 'excludedNamespaces' stuff seems super complex.  Does it need to be at this layer?  why not just have this layer store all the items, and then just filter at the completion layer?

Reply:
The main intention is to reduce the process time of the first invocation, since we likely don't need to traverse entire symbol trees with some namespaces in-scope (and possibly reduce cache size) However, this might not make a big enough difference to worth the extra complexity.

Let me do some profiling and decide if we should simplify this as you suggested.

Reply:
As mentioned offline, I was half way through the simplification process, but it turns out it's not trivial to support internalsVisibleTo if we just grab and cache everything. Additional logic is required and we ended up creating 2 different items for each internal type, one for projectswith internalsVisibleTo, another for projects without. As a result, it really doesn't simplify things much... Now I'm more inclined to keep the cache implementation as-is, unless you have a better reason to convince me not to

Reply:
Done